plugin/minimal/minimal.go

```diff
@@ -0,0 +1,41 @@
+package minimal
+
+import (
+	"context"
+	"time"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/nonwriter"
+	"github.com/coredns/coredns/plugin/pkg/response"
+	"github.com/miekg/dns"
+)
+
+// minimalResponse implements the plugin.Handler interface.
+type minimalResponse struct {
+	Next plugin.Handler
+}
+
+func (m *minimalResponse) Name() string { return "minimal-responses" }
+
+func (m *minimalResponse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	// if plugin active, try minimizing response
+	nw := nonwriter.New(w)
+
+	// Call all the next plugin in the chain.
+	rcode, err := plugin.NextOrFailure(m.Name(), m.Next, ctx, nw, r)
+	if err != nil {
+		// if error received then just return the error
+		return rcode, err
+	}
+	// else write minimized response
+	w.WriteMsg(m.minimizeResponse(nw.Msg))
+	return rcode, err
+}
+
+func (m *minimalResponse) minimizeResponse(msg *dns.Msg) *dns.Msg {
```

Comment:
so yeah, this is the thing that needs fleshing out

Reply:
Is this logic enough or is there anything else needed?

Reply:
This is a diff of what needs added here
~~~
diff --git plugin/minimal/README.md plugin/minimal/README.md
index 1781fb3d..165bb84e 100644
--- plugin/minimal/README.md
+++ plugin/minimal/README.md
@@ -6,8 +6,12 @@
 
 ## Description
 
-The *minimal* plugin tries to minimize the size of the response. Depending on the response type it removes resource records from the AUTHORITY and ADDITIONAL sections.
+The *minimal* plugin tries to minimize the size of the response. Depending on the response type it
+removes resource records from the AUTHORITY and ADDITIONAL sections.
 
+Specifically this plugin looks at successful responses (this excludes negative responses, i.e.
+nodata or name error). If the successful response isn't a delegation only the RRs in the answer
+section are written to the client.
 
 ## Syntax
 
@@ -21,7 +25,7 @@ Enable minimal responses:
 
 ~~~ corefile
 example.org {
-    whoami
+    forward . 8.8.8.8
     minimal
 }
 ~~~
diff --git plugin/minimal/minimal.go plugin/minimal/minimal.go
index c63cdeb0..bdd82019 100644
--- plugin/minimal/minimal.go
+++ plugin/minimal/minimal.go
@@ -2,11 +2,13 @@ package minimal
 
 import (
 	"context"
+	"fmt"
 	"time"
 
 	"github.com/coredns/coredns/plugin"
 	"github.com/coredns/coredns/plugin/pkg/nonwriter"
 	"github.com/coredns/coredns/plugin/pkg/response"
+
 	"github.com/miekg/dns"
 )
 
@@ -24,14 +26,34 @@ func (m *minimalHandler) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *
 	if err != nil {
 		return rcode, err
 	}
-	w.WriteMsg(m.minimizeResponse(nw.Msg))
-	return rcode, err
-}
+	// Depending on the type of response we can remove RRs from auth or additional section. Create a new message and
+	// copy the RRs we need into it.
+	ty, _ := response.Typify(nw.Msg, time.Now().UTC())
+	cl := response.Classify(ty)
+	// Denial or Error of left through, note this in includes Type=Meta and Type=Update, but that fine for  now
+	if cl == response.Denial || cl == response.Error {
+		w.WriteMsg(nw.Msg)
+		return 0, nil
+	}
+
+	// Further look at the response. We leave delegations alone (although we may prune those, but lets make it too difficult for ourselves
+	if ty == response.Delegation {
+		w.WriteMsg(nw.Msg)
+		return 0, nil
+	}
 
-func (m *minimalHandler) minimizeResponse(msg *dns.Msg) *dns.Msg {
-	if ty, _ := response.Typify(msg, time.Now().UTC()); ty == response.NoError {
-		msg.Extra = nil
-		msg.Ns = nil
+	if ty != response.NoError {
+		w.WriteMsg(nw.Msg)
+		return 0, plugin.Error("minimal", fmt.Errorf("unhandled response type %q for %q", ty, nw.Msg.Question[0].Name))
+	}
+	d := new(dns.Msg)
+	d.SetReply(r)
+	// Only copy the RRs in the answer section and write that to the client
+	d.Answer = make([]dns.RR, len(nw.Msg.Answer))
+	for i := range nw.Msg.Answer {
+		d.Answer[i] = nw.Msg.Answer[i] // this doesn't need to be a deep copy, we're not changing the RRs.
 	}
-	return msg
+
+	w.WriteMsg(d)
+	return 0, nil
 }
~~~

Reply:
As we don't need a deep copy can't we just set `d.Answer = nw.Msg.Answer`

Reply:
Also shouldn't the other parameters from the original msg be copied over too like header?

Reply:
I have pushed the suggested changes.